FIX Change error message of check_X_y when y is None

ClaudioSalvatoreArcidiacono opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs

Fixes #22576 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

Changes the error message of the `ValidationError` that is raised by `check_X_y` when y is None from:
```
y cannot be None
```
to
```
Expected array-like (array or non-string sequence), got None
```


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Change error message and tests

Commit:
Update Changelog

Commit:
Run black

thomasjpfan commented on sklearn/utils/validation.py:
```diff
@@ -1057,7 +1057,7 @@ def check_X_y(
     """
     if y is None:
-        raise ValueError("y cannot be None")
+        raise ValueError("Expected array-like (array or non-string sequence), got None")
```
I think including `y` in the error message is more informative. To pass the check we can do:

```python
    if y is None:
        if estimator is None:
            estimator_name = "estimator"
        else:
            estimator_name = _check_estimator_name(estimator)
        raise ValueError(
            f"{estimator_name} requires y to be passed, but the target y is None"
        )
```

ClaudioSalvatoreArcidiacono replied:
I agree, I changed the value error accordingly to your suggestion. Thanks!

thomasjpfan commented on doc/whats_new/v1.0.rst:
```diff
@@ -140,6 +140,11 @@ Changelog
   descriptions in the generated HTML. :pr:`21493` by
   :user:`Aurélien Geron <ageron>`.
 
+- |Fix| Changes the error message of the ValidationError raised by
+  :func:`utils.validation.check_X_y` when y is None so that it is compatible
+  with :func:`utils.estimator_checks.check_requires_y_none`. :pr:`22578` by
+  :user:`Claudio Salvatore Arcidiacono <ClaudioSalvatoreArcidiacono>`.
```
Can we move this into `doc/whats_new/v1.1.rst` ?

ClaudioSalvatoreArcidiacono replied:
Sure, it should have been moved now :)

thomasjpfan reviewed (commented):
Thank you for the PR! Can you add a non-regression test to `sklearn/utils/tests/test_estimator_checks.py` that contains your snippet from https://github.com/scikit-learn/scikit-learn/issues/22576 and assert no warning is raised?

An example of checking for no warnings is here:

https://github.com/scikit-learn/scikit-learn/blob/c36f926b66d22f514d5e87cc7f7b858ca026c5c7/sklearn/utils/tests/test_estimator_checks.py#L991-L994

Commit:
Move Changelog entry from v1.0 to v1.1

Commit:
Change error message accordingly to review comment

Commit:
Add test_check_requires_y_none

This test ensures consistency between the error returned by the
check_X_y function and the response that is checked by  the
check_requires_y_none function

Commit:
Refactor test_check_requires_y_none

ClaudioSalvatoreArcidiacono commented:
> Thank you for the PR! Can you add a non-regression test to `sklearn/utils/tests/test_estimator_checks.py` that contains your snippet from #22576 and assert no warning is raised?
> 
> An example of checking for no warnings is here:
> 
> https://github.com/scikit-learn/scikit-learn/blob/c36f926b66d22f514d5e87cc7f7b858ca026c5c7/sklearn/utils/tests/test_estimator_checks.py#L991-L994

Thanks for your suggestion. I have added an extra test to `sklearn/utils/tests/test_estimator_checks.py`.

thomasjpfan reviewed (approved):
Small nit, otherwise LGTM

ClaudioSalvatoreArcidiacono committed:
Update sklearn/utils/tests/test_estimator_checks.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Commit:
Remove line not covered in tests

jjerphan commented on doc/whats_new/v1.1.rst:
```diff
@@ -756,6 +756,11 @@ Changelog
   `check_methods_sample_order_invariance` and `check_methods_subset_invariance` tests.
   :pr:`22318` by `Zhehao Liu <MaxwellLZH>`.
 
+- |Fix| Changes the error message of the ValidationError raised by
+  :func:`utils.validation.check_X_y` when y is None so that it is compatible
+  with :func:`utils.estimator_checks.check_requires_y_none`. :pr:`22578` by
+  :user:`Claudio Salvatore Arcidiacono <ClaudioSalvatoreArcidiacono>`.
```
```suggestion
- |Fix| Changes the error message of the `ValidationError` raised by
  :func:`utils.validation.check_X_y` when y is None so that it is compatible
  with :func:`utils.estimator_checks.check_requires_y_none`. :pr:`22578` by
  :user:`Claudio Salvatore Arcidiacono <ClaudioSalvatoreArcidiacono>`.
```

jjerphan committed:
DOC Format

jjerphan committed:
Merge branch 'main' into 22576_change_error_message_check_X_y_when_y_is_None

thomasjpfan committed:
STY Fix
